### ISSUE TITLE
Figure out what media type and media properties to use for NG serialization format

### ISSUE THREAD

**Issue by BeeWarloc** [okb/Pomona#66]
What should the media type for the new pomona json serialization format be?

My proposal is `application/json;charset=utf-8;pomona.format=2`. Any input on this?

**Comment by asbjornu** [comment]
As recommended in [The RESTful Cookbok](http://restcookbook.com/Resources/using-custom-content-types/), I think we should define something like `application/vnd.bizi-storefront.customer+json; version=X`. That is both human readable, generic (`json`), specifies the type (`customer`) and allows for per-resource versioning (`version=X`).

**Comment by BeeWarloc** [comment]
Do you think having the entity as part of the media type important? A problem with that is knowing what to do in case of ambiguity between media type and `$type` property of content..

**Comment by asbjornu** [comment]
I'm thinking that whenever `type` is necessary, it should be the HTTP `Content-Type` and not the internal, C# specific `System.Type`. There should be a convention-based mapping between the two, but the only one exposed over HTTP should imho be an RFC 2046 MIME Media Type.

**Comment by BeeWarloc** [comment]
It wouldn't directly be the C# specific System.Type. There's already a type mapping set up, and that would be used regardless of whether the entity type is specified in the media type, or embedded as a part of the document.

My opinion on this is not written in stone, but I can't see any practical reasons why it would be better to make the entity type a part of the content-type header. I do however see some arguments about why not to do it:

* It's doesn't seem to be common. Although the `application/vnd.*+json` style seems to be spreading, most "meta formats" have chosen to include the entity type as a part of the body. OData, JSON-LD, JSON API. 
* A nested resource in a request or response do not have its own header.
* The serializer and deserializer in Pomona currently works with embedded type. Changing that would take extra effort.
* By keeping the type in the body the body can be cached as JSON in a document DB by itself without keeping the headers along with it.

**Comment by BeeWarloc** [comment]
There's also another consideration: inheritance.
To keep backwards compatibility I'm also thinking it might be a good idea to make the type specifier an array instead of a string.

Take the resource type `Payment`. At a later point another type `BitcoinPayment` is introduced which inherits `Payment`.

It is clear that `{ "$type": "BitcoinPayment" }` would not be recognizable by a legacy client. With `{ "$type": [ "BitcoinPayment", "Payment" ] }` the client could fall back to deserializing it as a `Payment`.

**Comment by asbjornu** [comment]
I'm not argumenting for the removal of `type` from the HTTP body. I'm just argumenting for using a full MIME Media Type wherever `type` in any form is necessary, be it inside a JSON document, within HTML and JavaScript documents referring to Pomona content or in the HTTP `Content-Type` header served by the Pomona backend.

I further argue that having the entity type inside the media type improves legibility, compatibility (with other clients than the autogenerated C# one), caching and serialization. Not knowing what the HTTP body is (and can be deserialized into) before it is downloaded and parsed into JSON makes HTTP methods like `HEAD` less useful, for instance.

I see your point with regards to inheritance, but using an opaque `System.Type`-like string or a full media type would not make much difference to that scenario. The interesting thing to note there is the array syntax to indicate the inheritance tree, which could just as well be expressed using media types:
```
{ type: [ "application/vnd.my-webshop.bitcoinpayment+json", "application/vnd.my-webshop.payment+json" ] }
```
It is also possible to expand the MIME type with parameters to achieve the same thing and avoid the array altogether:
```
{ type: "application/vnd.my-webshop.bitcoinpayment+json; entity=payment" }
```

**Comment by kedano** [comment]
The JSON API standardization might be of interest http://jsonapi.org



—
Kenneth Dahlstrøm

**Comment by BeeWarloc** [comment]
@asbjornu, I find using the full media type inside the document awkward and too verbose.

Could you describe a specific real life scenario where this would be useful?

**Comment by BeeWarloc** [comment]
@kedano, thanks. I did read through that earlier, but I prefer not splitting the document into `links` and `attributes`. Although I do admit there are certain advantages by doing that.

**Comment by BeeWarloc** [comment]
https://github.com/okb/Pomona/wiki/New-JSON-format

**Comment by GeirGrusom** [comment]
One advantage of specifying deserialization types in HTTP headers would be that the consumer could select a fitting parser before starting to read the document. The document can then be properly deserialized in one sitting.

**Comment by asbjornu** [comment]
I agree, @GeirGrusom.

@BeeWarloc wrote:
> I find using the full media type inside the document awkward and too verbose.
> Could you describe a specific real life scenario where this would be useful?

Yes, just having one way to indicate "type" in the API makes it more consistent and easier to understand. Having several "type" concepts, some with proprietary format and internal (obscure) usage, when HTTP has a standard for this, seems weird, imho.

**Comment by BeeWarloc** [comment]
@asbjornu,

I still think it looks really ugly, to embed the full media type in the document. In this context we already know that we're parsing JSON, and repeating that feels very redundant.

Could you to find some real life example of API where a similar design is used?

**Comment by asbjornu** [comment]
I completely agree that everything related to "Multipurpose Internet Mail Extensions" is ugly, but that's the standard we have, unfortunately. Having different "type" concepts in the API is a bad idea, imho.

**Comment by BeeWarloc** [comment]
Also the `$type` should ideally resolve to an URI with metadata. I have not figured out exactly how we should do this yet, but it's important for supporting references across API boundaries.

OData way of doing this:
http://docs.oasis-open.org/odata/odata-json-format/v4.0/errata02/os/odata-json-format-v4.0-errata02-os-complete.html#_Toc403940617

JSON schema way:
http://json-schema.org/latest/json-schema-core.html#anchor25

LD-JSON also does something similar. `$type` is not the same as media type. We usually don't describe the type of content in a webpage either, you don't do `text/vnd.bitbear.pr0n+html`. ;)

**Comment by BeeWarloc** [comment]
Btw, keeping the media type limited to describing the "outer" format is in no way against any REST principles or standards. It is perfectly viable, and seems to be the most usual way of doing things.

**Comment by BeeWarloc** [comment]
@asbjornu 

I got one more point to make here. I think the `$type` should be the same regardless of representation. The media type specifies the representation, but that could vary depending on the `Accept` header provided.

Currently there's no support for other serialization formats than JSON in pomona, but in the future we might want support for MessagePack or XML. If media type is provided in `$type` that would make it differ between the serialization formats.

**Comment by BeeWarloc** [comment]
@GeirGrusom  it's true that deserializing large JSON documents can be memory consuming when `$type` is not the first property provided.

OData solves this by appending a `odata.streaming=true` parameter to the media type, which adds the constraint that all "meta" properties has to be at the beginning of a JSON object. If there's no `odata.streaming=true` parameter available the properties have to be sorted, which means the whole body needs to be loaded in memory.

That could be a possibility in Pomona too, but my plan is to avoid serving large documents with 1+N nested resources, and instead provide the expanded resources in a stream. Then it would be the responsibility of the client to resolve references. That would make efficient caching easier.

**Comment by asbjornu** [comment]
@BeeWarloc You're starting to convince me. :wink: Especially by using URIs as a scheme for identifying `type`, which I think is a good idea. It's the same standard defined for `rel` in [Atom](https://tools.ietf.org/html/rfc4287#section-4.2.7.2) for relations that aren't IANA registered.

Have you thought about this in relation to the set of `rel` identifiers that should be exposed in the [Link header](https://tools.ietf.org/html/rfc5988) on `HEAD` request to `/` to give a list of the root resources available in the API? Perhaps these `rel` identifiers are the same as `$type`? Link header example:

```
Link: </critters>; rel="http://example.com/rel/critters"; title="Critters",
      </guns>; rel="http://example.com/rel/guns"; title="The different guns critters can equip"
```

**Comment by BeeWarloc** [comment]
@asbjornu excellent! :+1: 

Link headers.. To be honest I don't completely get it?
I think I sort of understand what they are, but in what specific scenario would I use them? As opposed to the ones that are part of the body?

**Comment by asbjornu** [comment]
In essence, it doesn't need to represent anything other than what's already within the HTTP body, but it can. See [this blog post by Mark Nottingham](https://www.mnot.net/blog/2012/10/29/NO_OPTIONS) for examples. Further, it's (much) cheaper to perform a `HEAD` request than a `GET` request.

**Comment by BeeWarloc** [comment]
When considering performance the serialization of the JSON body is a small part of the serialization process. If HEAD were to include link headers the entity would have to be loaded anyway, so I think the performance difference between HEAD and GET in that case would be negligible.

Why should links be a part of a HEAD response, but not other values? Semantically I see references as part of the content, not meta data?

What kind of adoption has link headers gotten btw?

**Comment by GeirGrusom** [comment]
`Link: </schemas/critters>; rel="type"`?

**Comment by GeirGrusom** [comment]
According to HTTP 1.1 a HEAD response should be identical to a bodyless GET from the same resource, so the Link header should be included in both GET and HEAD requests.

@BeeWarloc I don't think serialization is important here. It's more about what the client needs, and if they just want to know how to deserialize critters (since links are cachable) it should recieve the body anyway. This could used to build a parser.

**Comment by asbjornu** [comment]
@BeeWarloc, you need to consider the work required to query and process the data to deliver the JSON body too. Accumulated, this repeated querying, processing and serialization might not be negligible. Plus, as I mentioned, the `Link` header can contain (much) more metadata than what would be meaningful to put inside the body. I have no idea what adoption rate the header has; I have no means to figure that out. Ideas?

@GeirGrusom Yes, using `/schemas/` as a reserved root resource for everything schema related might be a good idea. And [`rel="type"`](http://tools.ietf.org/html/rfc6903#section-6), if I'm not reading it like a drunk sailor, seems to be a perfect match for what we're after here too!

**Comment by asbjornu** [comment]
Another option would be to use the "profile" link relation.

**Comment by asbjornu** [comment]
Haven't we agreed verbally to use `application/ld+json` as the next serialization format slash content type? If so, we should close this and create a new issue for implementing JSON-LD for version 2.0.
